### PR TITLE
feat: port `snapshot_from_lockfile` from denoland/deno

### PR DESCRIPTION
**Pull request by magurotuna**
This commit brings `snapshot_from_lockfile` function to `deno_npm` from the main `deno` repository. 

https://github.com/denoland/deno/pull/20024 shows how the caller side will change due to this port.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -23,3 +24,6 @@ futures = "0.3.28"
 pretty_assertions = "1.0.0"
 serde_json = { version = "1.0.67", features = ["preserve_order"] }
 tokio = { version = "1.27.0", features = ["full"] }
+
+[patch.crates-io]
+deno_lockfile = { git = "https://github.com/magurotuna/deno_lockfile.git", rev = "71d60975b89193ce2b8a68ebaee47882bb602afa" }
```
Comment (magurotuna) [inline Cargo.toml:29]: see https://github.com/denoland/deno_lockfile/pull/3

File: Cargo.lock [hunk 2]
```diff
@@ -382,6 +418,21 @@ dependencies = [
...
+version = "0.16.20"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "3053cf52e236a3ed746dfc745aa9cacf1b791d846bdaf412f60a8d7d6e17c8fc"
+dependencies = [
+ "cc",
+ "libc",
+ "once_cell",
+ "spin",
+ "untrusted",
+ "web-sys",
```
Comment (dsherret) [inline Cargo.lock:432]: I don't get why they have web-sys here.

File: src/resolution/snapshot.rs [hunk 3]
```diff
@@ -803,11 +808,154 @@ fn name_without_path(name: &str) -> &str {
...
+          return Err(err);
+        }
+      }
+    }
+
+    i += 1;
+  }
+
+  // clear the cache to reduce space for cache
+  api.clear_cache();
```
Comment (dsherret) [inline src/resolution/snapshot.rs:940]: I think we can remove this from here and instead call it after running this function in the CLI

File: src/registry.rs [hunk 4]
```diff
@@ -266,8 +266,52 @@ pub enum NpmRegistryPackageInfoLoadError {
   LoadError(#[from] Arc<anyhow::Error>),
 }
 
+/// A trait for getting package information from the npm registry, with no
+/// internal caching mechanism.
+#[async_trait]
+pub trait NpmRegistryApiWithoutCache: Sync + Send {
```
Comment (dsherret) [inline src/registry.rs:272]: Perhaps we could just have `NpmRegistryApi` to have default implementations that don't offer caching so we don't need an extra trait?

File: src/resolution/snapshot.rs [hunk 5]
```diff
@@ -803,11 +808,154 @@ fn name_without_path(name: &str) -> &str {
...
+}
+
+/// Constructs [`IncompleteSnapshot`] from the given [`Lockfile`]. The returned
+/// snapshot will then be passed to [`snapshot_from_lockfile`] to get a completely
+/// resolved snapshot.
+///
+/// The reason why this function is not combined with [`snapshot_from_lockfile`]
+/// is because we want `lockfile` to not live across the `.await` points. This
+/// becomes problematic if `lockfile` is wrapped in [`std::sync::Mutex`] or
+/// similar, which doesn't implement [`Send`].
```
Comment (dsherret) [inline src/resolution/snapshot.rs:842]: 👍

**Review by dsherret — COMMENTED** [review]
Looks almost good to me. Just a few small tweaks and this should be good.

File: Cargo.lock (hunk 2 above)
Comment (magurotuna) [reply Cargo.lock:465]: So it seems like deno_lockfile depends on ring, which in turn depends on web-sys (but this isn't included in the compilation, just included in the lock file). https://github.com/briansmith/ring/pull/904 is talking about it

File: src/registry.rs (hunk 4 above)
Comment (magurotuna) [reply src/registry.rs]: Good idea!

File: src/resolution/snapshot.rs (hunk 3 above)
Comment (magurotuna) [reply src/resolution/snapshot.rs]: Sounds good, now `clear_cache` method is removed from the trait, and it's moved to the CLI code: https://github.com/denoland/deno/pull/20024/files#diff-747e5bf0658bef1ca03cfa6b83c21b7ffc899ce4324557b2cc94356681e76699R841

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**